File: sdk/typescript/src/providers/google.ts
import type {
  ChatMessage,
  AnalysisResponse,
  MultimodalContentPart,
} from "../types.js";
import type {
  ProviderConfig,
  GoogleRequestBody,
  GoogleResponse,
  GoogleContentPart,
  ResponseFormat,
} from "./types.js";

/**
 * Remove additionalProperties from schema (Google API doesn't support it)
 */
function stripAdditionalProperties(
  schema: Record<string, unknown>
): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(schema)) {
    if (key === "additionalProperties") continue;
    if (typeof value === "object" && value !== null && !Array.isArray(value)) {
      result[key] = stripAdditionalProperties(value as Record<string, unknown>);
    } else {
      result[key] = value;
    }
  }
  return result;
}

/**
 * Convert multimodal content parts to Google format
 */
function toGoogleParts(
  content: string | MultimodalContentPart[]
): GoogleContentPart[] {
  if (typeof content === "string") {
    return [{ text: content }];
  }

  return content.map((part) => {
    if (part.type === "text") {
      return { text: part.text };
    }
    // Extract base64 data and media type from data URL
    const url = part.image_url.url;
    const match = url.match(/^data:([^;]+);base64,(.+)$/);
    if (match) {
      return {
        inlineData: {
          mimeType: match[1],
          data: match[2],
        },
      };
    }
    // If it's a regular URL, we can't use it directly with Google
    // They require inlineData for images
    throw new Error("Google requires base64 encoded images, not URLs");
  });
}

/**
 * Google Generative AI (Gemini) provider configuration
 * Supports structured outputs via responseSchema in generationConfig
 */
export const googleProvider: ProviderConfig = {
  baseUrl: "https://generativelanguage.googleapis.com/v1beta/models",
  envVar: "GOOGLE_API_KEY",

  authHeader: (apiKey: string) => ({
    "x-goog-api-key": apiKey,
    "Content-Type": "application/json",
  }),

  buildUrl: (baseUrl: string, model: string) =>
    `${baseUrl}/${model}:generateContent`,

  transformRequest: (
    model: string,
    messages: ChatMessage[],
    responseFormat?: ResponseFormat
  ): GoogleRequestBody => {
    // Extract system message if present
    const systemMessage = messages.find((m) => m.role === "system");
    const nonSystemMessages = messages.filter((m) => m.role !== "system");

    const requestBody: GoogleRequestBody = {
      contents: nonSystemMessages.map((m) => ({
        role: m.role === "assistant" ? "model" : "user",
        parts: toGoogleParts(m.content),
      })),
    };

    // Add system instruction if present (must be text only)
    if (systemMessage) {
      const systemContent =
        typeof systemMessage.content === "string"
          ? systemMessage.content
          : systemMessage.content
              .filter((p) => p.type === "text")
              .map((p) => (p as { type: "text"; text: string }).text)
              .join("\n");
      requestBody.systemInstruction = {
        parts: [{ text: systemContent }],
      };
    }

    // Transform response format to Google's generationConfig format
    // Strip additionalProperties as Google API doesn't support it
    if (responseFormat?.type === "json_schema") {
      requestBody.generationConfig = {
        responseMimeType: "application/json",
        responseSchema: stripAdditionalProperties(
          responseFormat.json_schema.schema as Record<string, unknown>
        ),
      };
    } else if (responseFormat?.type === "json_object") {
      requestBody.generationConfig = {
        responseMimeType: "application/json",
      };
    }

    return requestBody;
  },

  transformResponse: (response: unknown): AnalysisResponse => {
    const res = response as GoogleResponse;
    const candidate = res.candidates?.[0];
    const textContent = candidate?.content?.parts?.[0]?.text ?? "";

    return {
      id: res.modelVersion ?? "google-response",
      usage: {
        promptTokens: res.usageMetadata?.promptTokenCount ?? 0,
        completionTokens: res.usageMetadata?.candidatesTokenCount ?? 0,
        totalTokens: res.usageMetadata?.totalTokenCount ?? 0,
      },
      choices: [
        {
          index: 0,
          message: {
            role: "assistant",
            content: textContent,
          },
          finish_reason: candidate?.finishReason ?? "stop",
        },
      ],
    };
  },
};

File: sdk/python/src/safety_agent/providers/google.py
"""
Google (Gemini) provider configuration
"""

from typing import Any

from ..types import ChatMessage, AnalysisResponse, TokenUsage, AnalysisResponseChoice
from ..schemas import ResponseFormat


def _to_google_content(
    content: str | list[dict[str, Any]],
) -> list[dict[str, Any]]:
    """Convert multimodal content parts to Google format."""
    if isinstance(content, str):
        return [{"text": content}]

    result = []
    for part in content:
        if part.get("type") == "text":
            result.append({"text": part["text"]})
        elif part.get("type") == "image_url":
            # Extract base64 data from data URL
            url = part["image_url"]["url"]
            if url.startswith("data:"):
                # Parse data URL: data:image/png;base64,xxxxx
                parts = url.split(",", 1)
                if len(parts) == 2:
                    mime_type = parts[0].split(":")[1].split(";")[0]
                    data = parts[1]
                    result.append({
                        "inlineData": {
                            "mimeType": mime_type,
                            "data": data,
                        }
                    })
    return result


class GoogleProvider:
    """Google (Gemini) provider configuration."""

    base_url = "https://generativelanguage.googleapis.com/v1beta/models"
    env_var = "GOOGLE_API_KEY"

    def auth_header(self, api_key: str) -> dict[str, str]:
        return {
            "Content-Type": "application/json",
        }

    def build_url(self, base_url: str, model: str) -> str:
        import os
        api_key = os.environ.get(self.env_var, "")
        return f"{base_url}/{model}:generateContent?key={api_key}"

    def transform_request(
        self,
        model: str,
        messages: list[ChatMessage],
        response_format: ResponseFormat | None = None,
    ) -> dict[str, Any]:
        # Separate system message from others
        system_parts = []
        contents = []

        for msg in messages:
            if msg.role == "system":
                system_parts.append({"text": msg.content if isinstance(msg.content, str) else ""})
            else:
                role = "model" if msg.role == "assistant" else "user"
                contents.append({
                    "role": role,
                    "parts": _to_google_content(msg.content),
                })

        request: dict[str, Any] = {
            "contents": contents,
        }

        if system_parts:
            request["systemInstruction"] = {"parts": system_parts}

        # Add structured output format if provided
        if response_format and response_format.get("type") == "json_schema":
            json_schema = response_format.get("json_schema", {})
            request["generationConfig"] = {
                "responseMimeType": "application/json",
                "responseSchema": json_schema.get("schema"),
            }

        return request

    def transform_response(self, response: dict[str, Any]) -> AnalysisResponse:
        content = ""
        candidates = response.get("candidates", [])
        if candidates:
            candidate = candidates[0]
            candidate_content = candidate.get("content", {})
            for part in candidate_content.get("parts", []):
                if "text" in part:
                    content += part["text"]

        usage = response.get("usageMetadata", {})
        prompt_tokens = usage.get("promptTokenCount", 0)
        completion_tokens = usage.get("candidatesTokenCount", 0)

        return AnalysisResponse(
            id=response.get("modelVersion", ""),
            usage=TokenUsage(
                prompt_tokens=prompt_tokens,
                completion_tokens=completion_tokens,
                total_tokens=usage.get("totalTokenCount", prompt_tokens + completion_tokens),
            ),
            choices=[
                AnalysisResponseChoice(
                    index=0,
                    message=ChatMessage(role="assistant", content=content),
                    finish_reason=candidates[0].get("finishReason", "stop") if candidates else "stop",
                )
            ],
        )


google_provider = GoogleProvider()
